src/MICore/LaunchOptions.cs

```diff
@@ -83,7 +87,106 @@ public PipeLaunchOptions(string pipePath, string pipeArguments, string pipeComma
+            }
+
+            string pipeProgram = EnsurePipeProgramExists(pipeTransport);
+            if (string.IsNullOrWhiteSpace(pipeProgram))
+            {
+                throw new InvalidLaunchOptionsException(String.Format(CultureInfo.InvariantCulture, MICoreResources.Error_PipeProgramNotFound, pipeTransport.PipeProgram));
+            }
+
+            PipeLaunchOptions pipeOptions = new PipeLaunchOptions(
+                pipePath: pipeProgram,
+                pipeArguments: EnsurePipeArguments(pipeTransport.PipeArgs, pipeTransport.DebuggerPath, gdbPathDefault, pipeTransport.QuoteArgs.GetValueOrDefault(true)),
+                pipeCommandArguments: ParseArguments(pipeTransport.PipeArgs, pipeTransport.QuoteArgs.GetValueOrDefault(true)),
+                pipeCwd: pipeTransport.PipeCwd,
+                pipeEnvironment: GetEnvironmentEntries(pipeTransport.PipeEnv)
+                );
+
+            Json.LaunchOptions.BaseOptions baseOptions = Json.LaunchOptions.LaunchOptionHelpers.GetLaunchOrAttachOptions(parsedOptions);
+            pipeOptions.InitializeCommonOptions(baseOptions);
+            if (baseOptions is Json.LaunchOptions.LaunchOptions)
+            {
+                pipeOptions.InitializeLaunchOptions((Json.LaunchOptions.LaunchOptions)baseOptions);
+            }
+
+            if (baseOptions is Json.LaunchOptions.AttachOptions)
+            {
+                pipeOptions.InitializeAttachOptions((Json.LaunchOptions.AttachOptions)baseOptions);
+            }
+
+            return pipeOptions;
+        }
+
+        private static string EnsurePipeProgramExists(Json.LaunchOptions.PipeTransport pipeTransport)
+        {
+            string pipeProgram = pipeTransport.PipeProgram;
+            // PipeProgram should exist, if not, concatinate it with the pipeCwd
+            if (File.Exists(pipeProgram))
+            {
+                return pipeProgram;
+            }
+
+            if (!String.IsNullOrWhiteSpace(pipeTransport.PipeCwd))
+            {
+                pipeProgram = Path.Combine(pipeTransport.PipeCwd, pipeTransport.PipeProgram);
+                if (File.Exists(pipeProgram))
+                {
+                    return pipeProgram;
+                }
+            }
+
+            return null;
+        }
+
+        private static string EnsurePipeArguments(List<string> pipeArgs, string debuggerPath, string debuggerPathDefault, bool quoteArgs)
+        {
+            // Debugger path. Assume /usr/bin/gdb unless specified
+            string dbgPath = String.IsNullOrWhiteSpace(debuggerPath) ? debuggerPathDefault : debuggerPath;
+
+            // debugger command: /usr/bin/gdb --interpreter=mi
+            string dbgCmdArguments = String.Format(CultureInfo.InvariantCulture, "{0} {1}", dbgPath, "--interpreter=mi");
+
+            string userArguments = ParseArguments(pipeArgs, quoteArgs);
+
+            if (userArguments.Contains("${debuggerCommand}"))
```

Comment:
>if [](start = 12, length = 2)

We should probably extract this if/else block into a helper method that we can use both here and also at:

```
PipeTransport.cs(319):     string fullCommand = string.Format(CultureInfo.InvariantCulture, "{0} {1} {2}", _pipePath, _cmdArgs, commandText);
```

(Or just fix up that line also)